tool/common/update/update.go

```diff
@@ -0,0 +1,431 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package update
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/client/webclient"
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	teleportToolsVersion = "TELEPORT_TOOLS_VERSION"
+	checksumHexLen       = 64
+	reservedFreeDisk     = 10 * 1024 * 1024
+
+	FlagEnt  = 1 << 0
+	FlagFips = 1 << 1
+)
+
+var (
+	pattern       = regexp.MustCompile(`(?m)Teleport v(.*) git`)
+	baseUrl       = "https://cdn.teleport.dev"
+	defaultClient = newClient(&downloadConfig{})
+	featureFlag   int
+)
+
+// CheckLocal is run at client tool startup and will only perform local checks.
+func CheckLocal() (string, bool) {
+	// If a version of client tools has already been downloaded to
+	// $TELEPORT_HOME/bin, return that.
+	toolsVersion, err := version()
+	if err != nil {
+		return "", false
+	}
+
+	// Check if the user has requested a specific version of client tools.
+	requestedVersion := os.Getenv(teleportToolsVersion)
+	switch {
+	// The user has turned off any form of automatic updates.
+	case requestedVersion == "off":
+		return "", false
+	// Requested version already the same as client version.
+	case teleport.Version == requestedVersion:
+		return requestedVersion, false
+	// The user has requested a specific version of client tools.
+	case requestedVersion != "" && requestedVersion != toolsVersion:
+		return requestedVersion, true
+	}
+
+	return toolsVersion, false
```

Comment:
@russjones If the user downloads newer version of tsh/tctl and let say execute it from download directory, should we alway re-execute with saved version from tools directory (where version might be older)